sys/arm64/include/cherireg.h

```diff
@@ -0,0 +1,167 @@
+ * CHERI_PERMS_SWALL: Mask of all available software-defined permissions
+ * CHERI_PERMS_HWALL: Mask of all available hardware-defined permissions
+ */
+#define	CHERI_PERMS_SWALL						\
+	(CHERI_PERM_SW0 | CHERI_PERM_SW1 | CHERI_PERM_SW2 |		\
+	CHERI_PERM_SW3)
+
+#define	CHERI_PERMS_HWALL						\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_EXECUTIVE |			\
+	CHERI_PERM_MUTABLE_LOAD | CHERI_PERM_COMPARTMENT_ID |		\
+	CHERI_PERM_BRANCH_UNSEAL | CHERI_PERM_SYSTEM |			\
+	CHERI_PERM_UNSEAL | CHERI_PERM_SEAL | 				\
+	CHERI_PERM_STORE_LOCAL_CAP | CHERI_PERM_STORE_CAP |		\
+	CHERI_PERM_LOAD_CAP | CHERI_PERM_EXECUTE | CHERI_PERM_STORE |	\
+	CHERI_PERM_LOAD)
+
+/*
+ * Root "object-type" capability for the kernel.  This can be used neither as
+ * a data nor code capability.
+ */
+#define	CHERI_PERM_KERN_TYPE	(CHERI_PERM_GLOBAL | CHERI_PERM_SEAL |	\
+	CHERI_PERM_UNSEAL)
+
+/*
+ * Basic userspace permission mask; CHERI_PERM_EXECUTE will be added for
+* executable capabilities (pcc); CHERI_PERM_STORE, CHERI_PERM_STORE_CAP,
+ * and CHERI_PERM_STORE_LOCAL_CAP will be added for data permissions (ddc).
+ */
+#define	CHERI_PERMS_USERSPACE						\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP |	\
+	(CHERI_PERMS_SWALL & ~CHERI_PERM_CHERIABI_VMMAP))
+
+#define	CHERI_PERMS_USERSPACE_CODE					\
+	(CHERI_PERMS_USERSPACE | CHERI_PERM_EXECUTE)
+
+#define	CHERI_PERMS_USERSPACE_SEALCAP					\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_SEAL | CHERI_PERM_UNSEAL)
+
+/*
+ * _DATA includes _VMMAP so that we can derive the MMAP cap from it.
+ *
+ * XXX: We may not want to include VMMAP here and instead only in
+ * CHERI_CAP_USER_MMAP_PERMS
+ */
+#define	CHERI_PERMS_USERSPACE_DATA					\
+				(CHERI_PERMS_USERSPACE |		\
+				CHERI_PERM_STORE |			\
+				CHERI_PERM_STORE_CAP |			\
+				CHERI_PERM_STORE_LOCAL_CAP |		\
+				CHERI_PERM_CHERIABI_VMMAP)
+
+/*
+ * Corresponding permission masks for kernel code and data; these are
+ * currently a bit broad, and should be narrowed over time as the kernel
+ * becomes more capability-aware.
+ */
+#define	CHERI_PERMS_KERNEL						\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP)	\
+
+#define	CHERI_PERMS_KERNEL_CODE						\
+	(CHERI_PERMS_KERNEL | CHERI_PERM_EXECUTE | CHERI_PERM_SYSTEM_REGS)
+
+#define	CHERI_PERMS_KERNEL_DATA						\
+	(CHERI_PERMS_KERNEL | CHERI_PERM_STORE | CHERI_PERM_STORE_CAP |	\
+	CHERI_PERM_STORE_LOCAL_CAP)
+
+#define	CHERI_PERMS_KERNEL_SEALCAP					\
+	(CHERI_PERM_GLOBAL | CHERI_PERM_SEAL | CHERI_PERM_UNSEAL)
+
+/*
+ * The CHERI object-type space is split between userspace and kernel,
+ * permitting kernel object references to be delegated to userspace (if
+ * desired).  Currently, we provide 13 bits of namespace to each, with the top
+ * bit set for kernel object types, but it is easy to imagine other splits.
+ * User and kernel software should be written so as to not place assumptions
+ * about the specific values used here, as they may change.
+ */
+#define	CHERI_OTYPE_BITS	(14)
+#define	CHERI_OTYPE_USER_MIN	(0)
```

Comment:
This should be 4, since the first 3 are reserved.
```suggestion
#define	CHERI_OTYPE_USER_MIN	(4)
```